### ISSUE TITLE
Perplexity higher than GPT2 paper

### ISSUE THREAD

**Issue by TheophileBlard** [antoiloui/belgpt2#1]
In the [GPT2 paper](https://d4mucfpksywv.cloudfront.net/better-language-models/language_models_are_unsupervised_multitask_learners.pdf), the authors report a perplexity of ~16 for GPT2-small trained on the WebText dataset. Several open-source projects in several languages also achieve a similar performance (perplexity around 20-30). In the README, we can see that the model doesn't go below 52 of perplexity, do you know what could be the reason of this ?

I tested your pretrained model available in the [HuggingFace Hub](https://huggingface.co/antoiloui/belgpt2), and observed that there is no correlation whatsoever between consecutive sentences. Each sentence on its own bears a meaning, but they do not work together, resulting in an incomprehensible piece of text. Here are some example of sentences I generated with the code you provide on the README:

> GPT2 est le programme de formation de la police fédérale et de police locale des Pays-Bas. Et comme les jeux d\' argent fonctionnent, ils attirent beaucoup de gens dans les casinos en ligne. Le nouveau Premier ministre, dont on sait qu\' il a été reçu par son homologue russe Vladimir Poutine, " avait déjà prévenu qu\' il fallait tenir compte de la situation sur le terrain et il y avait encore quelque chose d\' important en suspens. " Si vous ne voulez pas aller voir un

> GPT2 est de l ordre de 0,183, la différence avec la différence entre les valeurs ( et donc la masse ) de charges partielles, donc un gain à mesurer par rapport aux charges partielles, que l on prend comme valeur de l énergie cinétique. Une fois n' est pas coutume, ce sera une autre fois. Un petit côté pop, presque disco mais efficace. Les photos et les descriptions sont conformes à la réalité. C' est cette même dynamique, ce même désir de toujours

Could this explain why the perplexity is that high ?

**Comment by ant-louis** [comment]
Hey, I think that the reported perplexity depends strongly on the dataset on which it is calculated, and that it is difficult to draw conclusions by comparing perplexities computed on different datasets (what’s more, in different languages). For example if you look at the original GPT-2 paper, you can see that the small version reaches a perplexity of 35.13 on LAMBADA, but also obtains a perplexity of 65.85 on PTB and even 75.2 on 1BW. 

Hence, it could simply be that the test set on which I calculated these perplexities is complex for the task of language modeling. To find out, it would be ideal to calculate the perplexity of BelGPT-2 on other French datasets commonly used as benchmarks.

However, I have also noticed these inconsistencies when the model produces several consecutive sentences, and I don't have an explanation for this yet.

**Comment by TheophileBlard** [comment]
Thank you for your response.
You're right, we should not draw conclusions without comparing to other French models. However, I don't think you can say that a "complex" test set is the same thing than a completely other dataset, on which the model hasn't be finetuned on. Was the final perplexity on your train set similar to the test set ?

Unfortunately, there are not many large French datasets around, and I think you already used most of them in your train set... And there are not (yet) many generative model in French language to which we could compare yours. It could be nice to establish a baseline for this specific task in French!

Full disclaimer, I also wanted to train a GPT2 model from scratch, then finetune it to generate movie reviews, based on the [allocine dataset](https://huggingface.co/nlp/viewer/?dataset=allocine). But I have no way near the compute power you used for belGPT2, and I only manage to train on Wikipedia. Thereby, I was thinking about using your model instead, but those inconsistencies scare me.

EDIT: I forgot to say: it IS really nice work.
